program/src/processor/slash.rs

```diff
@@ -11,14 +27,219 @@ use crate::instruction::accounts::{Context, SlashAccounts};
+
+    let signer_bump = [config.vault_authority_bump];
+    let derivation = create_vault_pda(ctx.accounts.config.key, &signer_bump, program_id)?;
+
+    require!(
+        ctx.accounts.vault_authority.key == &derivation,
+        StakeError::InvalidAuthority,
+        "vault authority",
+    );
+
+    let signer_seeds = get_vault_pda_signer_seeds(ctx.accounts.config.key, &signer_bump);
+
+    // vault
+    // - must be the token account on the stake config account
+
+    require!(
+        ctx.accounts.vault.key == &config.vault,
+        StakeError::IncorrectVaultAccount,
+    );
+
+    let vault_data = ctx.accounts.vault.try_borrow_data()?;
+    // unpack to validate the mint
+    let vault = PodStateWithExtensions::<PodAccount>::unpack(&vault_data)?;
+
+    // mint
+    // - must match the stake vault mint
+
+    require!(
+        &vault.base.mint == ctx.accounts.mint.key,
+        StakeError::InvalidMint,
+        "mint"
+    );
+
+    // Update the stake amount on both stake and config accounts:
+    //
+    //   1. the amount slashed is taken from the stake amount (this includes
+    //      the amount that is currently staked + deactivating);
+    //
+    //   2. if not enough, the remaining is ignored and the stake account is
+    //      left with 0 amount;
+    //
+    //   3. if the stake account has deactivating tokens, make sure that the
+    //      deactivating amount is at least the same as the stake amount (it might
+    //      need to be reduced due to slashing).
+
+    require!(
+        amount > 0,
+        StakeError::InvalidAmount,
+        "amount must be greater than 0"
+    );
+
+    // slashes active stake
+    let active_stake_to_slash = min(amount, stake.amount);
+    stake.amount = stake
+        .amount
+        .checked_sub(active_stake_to_slash)
+        .ok_or(ProgramError::ArithmeticOverflow)?;
+
+    if stake.deactivating_amount > stake.amount {
+        // adjust the deactivating amount if it's greater than the "active" stake
+        // after slashing
+        stake.deactivating_amount = stake.amount;
+        // clear the deactivation timestamp if the stake amount is 0
+        if stake.amount == 0 {
+            stake.deactivation_timestamp = None;
+        }
+    }
+
+    // Update the token amount delegated on the config account.
+    //
+    // The instruction will fail if the amount to slash is greater than the
+    // total amount delegated (it should never happen).
+    require!(
+        config.token_amount_delegated >= active_stake_to_slash,
+        StakeError::InvalidSlashAmount,
+        "slash amount greater than total amount delegated ({} required, {} delegated)",
+        active_stake_to_slash,
+        config.token_amount_delegated
+    );
+
+    config.token_amount_delegated = config
+        .token_amount_delegated
+        .checked_sub(active_stake_to_slash)
+        .ok_or(ProgramError::ArithmeticOverflow)?;
+
+    if amount.saturating_sub(active_stake_to_slash) > 0 {
```

Comment:
nit: maybe this is just me, but I find `amount > active_stake_to_slash` to be clearer than saturating math